clean/new_orleans_pd_uof.py

```diff
@@ -0,0 +1,552 @@
+from lib.columns import clean_column_names
+import deba
+from lib.clean import clean_sexes, standardize_desc_cols
+from lib.columns import set_values
+from lib.clean import clean_races, clean_names
+from lib.uid import gen_uid
+import numpy as np
+import pandas as pd
+import sys
+
+sys.path.append("../")
+
+
+def join_citizen_columns(df):
+    citizen_columns = [
+        "citizen_race",
+        "citizen_sex",
+        "citizen_hospitalized",
+        "citizen_injured",
+        "citizen_influencing_factors",
+        "citizen_distance_from_officer",
+        "citizen_age",
+        "citizen_build",
+        "citizen_height",
+        "citizen_arrested",
+        "citizen_arrest_charges",
+    ]
+    for col in citizen_columns:
+        df.loc[:, col] = df[col].str.split(r" \| ")
+
+    def create_citizen_dicts(row: pd.Series):
+        d = row.loc[citizen_columns].loc[row.notna()].to_dict()
+        keys = d.keys()
+        return [
+            {k: d[k][i] for k in keys} for i in range(max(len(v) for v in d.values()))
+        ]
+
+    df.loc[:, "citizen"] = df.apply(create_citizen_dicts, axis=1)
+    df = df.drop(
+        columns=[
+            "citizen_race",
+            "citizen_sex",
+            "citizen_hospitalized",
+            "citizen_injured",
+            "citizen_influencing_factors",
+            "citizen_distance_from_officer",
+            "citizen_age",
+            "citizen_build",
+            "citizen_height",
+            "citizen_arrested",
+            "citizen_arrest_charges",
+        ]
+    )
+
+    df.loc[:, "citizen"] = df.citizen.apply(pd.DataFrame)
+    df = df.drop("citizen", axis=1).join(
+        pd.concat(df["citizen"].values, keys=df.index).droplevel(1)
+    )
```

Comment:
This is not very efficient for the following reasons:
- You're still creating dictionaries but if we're going for a DataFrame, creating lists is more efficient. We already have the column names in a list after all.
- You're constructing a data frame for each cell, and then concat them. `pd.DataFrame` do a lot of preprocessing such as indexing data upon creation so it is more efficient to concatenate all records into a big record list and then call `pd.DataFrame` only once.
- No need to join with all the existing columns. We only need to tack on the `tracking_number`.
```python
# return a list of records for each row, notice that we also prepend
# tracking_number to each record.
def create_citizen_lists(row: pd.Series):
    d = row.loc[citizen_columns].loc[row.notna()].to_dict()
    return [
        [row.tracking_number] + [d[k][i] for k in d.keys()]
        for i in range(max(len(v) for v in d.values()))
    ]

citizen_series = df.apply(create_citizen_lists, axis=1)
citizen_df = pd.DataFrame(
    # double list comprehension is the fastest way to concat a list of lists
    [element for list_ in citizen_series for element in list_],
    columns=['tracking_number']+citizen_columns,
)
```